reader-privacy.md

```diff
@@ -0,0 +1,242 @@
+
+### Threat Modelling
+
+There are three actors involved into the IPNI workflow: provider, client and indexer. Providers update index by publishing advertisements.
+Indexer advertisements are signed by their publishers and can be authenticated. Advertisements are organised in a chain and are ingested strictly in order. 
+It's not possible to reorder advertisements wihtout forking the chain. Advertisements processing is idempotent - re-ingesting the same advertismeent twice 
+shouldn't affect the indexer's state. The IPNI specification is agnostic to transport protocols so particular protocol choice is up to the implementation. 
+Compromised publisher's identity is out of scope of this specification.
+
+Clients consume index by performing CID lookups. This specification introduces additional hashing and encryption that aim to prevent a passive observer
+from being able to infer what data is being looked up by spying at the client to indexer traffic. The exact communicaton protocol is out of scope for this specification 
+however it should be chosen carefully to prevent MITM attacks. Before Writer Privacy upgrade a passive observer could deobfuscate the content by building a `hash(Multihash)` to `Multihash` map.
+That however will not be possible eventually as mentioned in the [Security](#security) section. 
+
+Provider records returned to client lookups do not contain any authentication data. It's possible for a malicious / buggy indexer to  
+present a wrong dataset to the client. Clients can tackle that problem by excluding such indexers from their pool. Returning wrong datasets will
+eventually affect the indexer's reputation score - these efforts are already in progress. Data integrity is inbuilt into IPFS - clients
+can verify that the data returned by a storage provider matches the CID. So even if indexer get compromised that will not compromise
+the data itself. 
+
+## Implementation
+
+This section goes through Reader Privacy protocol, hashing and cryptography choices as well as explains some less obvious details 
+for the implementations to consider.
+
+### Hashing
+
+SHA256 must be used as a hashing function. To avoid collisions between the hashes calculated from the same value but for different purposes (for example
+to derive an encryption key versus derive a lookup key) a constant string can be appended to the value before calculating a hash over it. For example
+`sha256("AESGCM" + multihash)` to derive an encryption key and `sha256("CR_DOUBLEHASH" + multihash)` to calculate a second hash over the multihash.
+
+All hashed data that is used for lookups must be of `Multihash` format with `SHA_256` codec. Double hashed data must use `DBL_SHA_256` codec.
+
+Multihashes must be prepended with `CR_DOUBLEHASH` before calculating a second hash. Unhashed data must be prepended with `CR_HASH` before calculating the first hash.
```

Comment:
Are you sure you're meaning to use the DBL_SHA_256 code? It looks like that code and implementation are already defined. e.g. https://github.com/multiformats/go-multihash/blob/608669da49b636a646de3472101d0183889ae6c4/core/errata.go#L29.

`SHA256("CR_DOUBLEHASH" || SHA256(CR_HASH, <data>) != SHA256(SHA256(<data>))` which are both different from `SHA256("CR_DOUBLEHASH" || SHA256-MH(CR_HASH, <data>)`

Reply:
This is to be aligned with the current [DHT implementation](https://github.com/ChainSafe/go-libp2p-kad-dht/blob/dc84e3ec9de47811c1d28d5b644cc34e4a6c4d53/internal/utils.go#L12). It uses `DBL_SHA_256` codec as well as prepends the bytes with `CR_DOUBLEHASH` before calculating a second multihash.

Reply:
@aschmahmann do you suggest defining a new constant string for the purpose of double hashing in both the DHT and Indexers?